Random failures on TestStatisticsOnChannelOut

TestStatisticsOnChannelOut is failing randomly yet.
```
cespinoza@chriss-MBP ~/g/s/g/t/device-hub> make test
go test -covermode=atomic github.com/thingful/device-hub github.com/thingful/device-hub/cmd/device-hub github.com/thingful/device-hub/cmd/device-hub-cli github.com/thingful/device-hub/describe github.com/thingful/device-hub/endpoint github.com/thingful/device-hub/engine github.com/thingful/device-hub/listener github.com/thingful/device-hub/mocks github.com/thingful/device-hub/proto github.com/thingful/device-hub/registry github.com/thingful/device-hub/runtime github.com/thingful/device-hub/server github.com/thingful/device-hub/store github.com/thingful/device-hub/test-configuration github.com/thingful/device-hub/utils github.com/thingful/device-hub/utils/testing
?   	github.com/thingful/device-hub	[no test files]
?   	github.com/thingful/device-hub/cmd/device-hub	[no test files]
?   	github.com/thingful/device-hub/cmd/device-hub-cli	[no test files]
ok  	github.com/thingful/device-hub/describe	0.027s	coverage: 67.9% of statements
?   	github.com/thingful/device-hub/endpoint	[no test files]
ok  	github.com/thingful/device-hub/engine	0.020s	coverage: 76.6% of statements
ok  	github.com/thingful/device-hub/listener	0.022s	coverage: 28.7% of statements
?   	github.com/thingful/device-hub/mocks	[no test files]
?   	github.com/thingful/device-hub/proto	[no test files]
ok  	github.com/thingful/device-hub/registry	0.028s	coverage: 42.9% of statements
--- FAIL: TestStatisticsOnChannelOut (0.00s)
        Error Trace:    loop_test.go:127
	Error:		Not equal: 0x1 (expected)
			        != 0x0 (actual)

        Error Trace:    loop_test.go:128
	Error:		Not equal: 0x1 (expected)
			        != 0x0 (actual)

FAIL
coverage: 23.2% of statements
FAIL	github.com/thingful/device-hub/runtime	0.024s
?   	github.com/thingful/device-hub/server	[no test files]
ok  	github.com/thingful/device-hub/store	0.024s	coverage: 58.5% of statements
ok  	github.com/thingful/device-hub/test-configuration	0.022s	coverage: 0.0% of statements
?   	github.com/thingful/device-hub/utils	[no test files]
?   	github.com/thingful/device-hub/utils/testing	[no test files]
make: *** [test] Error 1
```

--- a/mocks/mocks.go
+++ b/mocks/mocks.go
@@ -27,9 +27,15 @@ func (m *Channel) Close() error {
 // Endpoint allows to create a mock Endpoint
 // Implements hub.Endpoint
 type Endpoint struct {
-	Error error
+	Writer func(message hub.Message) error
+	Error  error
 }
 
 func (e *Endpoint) Write(message hub.Message) error {
+
+	if e.Writer != nil {
+		return e.Writer(message)
+	}
+
 	return e.Error
 }

--- a/runtime/loop_test.go
+++ b/runtime/loop_test.go
@@ -82,7 +82,8 @@ func TestStatisticsOnChannelOut(t *testing.T) {
 
 	ctx := context.Background()
 
-	ctx, closerFunc := context.WithCancel(ctx)
+	wg := sync.WaitGroup{}
+	wg.Add(3)
 
 	messageChannel := make(chan hub.Message)
 
@@ -90,7 +91,7 @@ func TestStatisticsOnChannelOut(t *testing.T) {
 		ErrorChannel:   make(chan error),
 		MessageChannel: messageChannel,
 		Closer: func() error {
-			closerFunc()
+			wg.Done()
 			return nil
 		},
 	}
@@ -98,9 +99,17 @@ func TestStatisticsOnChannelOut(t *testing.T) {
 	pipe := newRuntimePipe(store.Pipe{})
 
 	endpoints := map[string]hub.Endpoint{
-		"ok": &mocks.Endpoint{},
+		"ok": &mocks.Endpoint{
+			Writer: func(hub.Message) error {
+				defer wg.Done()
+				return nil
+			},
+		},
 		"error": &mocks.Endpoint{
-			Error: errors.New("boo!"),
+			Writer: func(hub.Message) error {
+				defer wg.Done()
+				return errors.New("boo")
+			},
 		},
 	}
 
@@ -113,8 +122,7 @@ func TestStatisticsOnChannelOut(t *testing.T) {
 	messageChannel <- message
 
 	mock.Close()
-
-	<-ctx.Done()
+	wg.Wait()
 
 	assert.Equal(t, uint64(1), pipe.Statistics.Processed.Ok)
 	assert.Equal(t, uint64(1), pipe.Statistics.Processed.Total)

--- a/runtime/loop_test.go
+++ b/runtime/loop_test.go
@@ -56,12 +56,19 @@ func TestStatisticsOnChannelError(t *testing.T) {
 	t.Parallel()
 
 	ctx := context.Background()
+	ctx, closerFunc := context.WithCancel(ctx)
+	wg := sync.WaitGroup{}
+	wg.Add(1)
 
 	errorChannel := make(chan error)
 
 	mock := &mocks.Channel{
 		ErrorChannel:   errorChannel,
 		MessageChannel: make(chan hub.Message),
+		Closer: func() error {
+			wg.Done()
+			return nil
+		},
 	}
 
 	pipe := newRuntimePipe(store.Pipe{})
@@ -69,6 +76,12 @@ func TestStatisticsOnChannelError(t *testing.T) {
 	go loop(ctx, pipe, nil, map[string]hub.Endpoint{}, mock, utils.NewNoOpLogger(), map[string]string{})
 
 	errorChannel <- errors.New("boo!")
+
+	// closing the context will close the channel
+	closerFunc()
+
+	<-ctx.Done()
+	wg.Wait()
 
 	assert.Equal(t, uint64(1), pipe.Statistics.Received.Total)
 	assert.Equal(t, uint64(1), pipe.Statistics.Received.Errors)
@@ -81,6 +94,7 @@ func TestStatisticsOnChannelOut(t *testing.T) {
 	t.Parallel()
 
 	ctx := context.Background()
+	ctx, closerFunc := context.WithCancel(ctx)
 
 	wg := sync.WaitGroup{}
 	wg.Add(3)
@@ -91,7 +105,7 @@ func TestStatisticsOnChannelOut(t *testing.T) {
 		ErrorChannel:   make(chan error),
 		MessageChannel: messageChannel,
 		Closer: func() error {
-			defer wg.Done()
+			wg.Done()
 			return nil
 		},
 	}
@@ -101,13 +115,13 @@ func TestStatisticsOnChannelOut(t *testing.T) {
 	endpoints := map[string]hub.Endpoint{
 		"ok": &mocks.Endpoint{
 			Writer: func(hub.Message) error {
-				defer wg.Done()
+				wg.Done()
 				return nil
 			},
 		},
 		"error": &mocks.Endpoint{
 			Writer: func(hub.Message) error {
-				defer wg.Done()
+				wg.Done()
 				return errors.New("boo")
 			},
 		},
@@ -121,7 +135,10 @@ func TestStatisticsOnChannelOut(t *testing.T) {
 	}
 	messageChannel <- message
 
-	mock.Close()
+	// closing the context will close the channel
+	closerFunc()
+	<-ctx.Done()
+
 	wg.Wait()
 
 	assert.Equal(t, uint64(1), pipe.Statistics.Processed.Ok)

--- a/runtime/loop_test.go
+++ b/runtime/loop_test.go
@@ -91,7 +91,7 @@ func TestStatisticsOnChannelOut(t *testing.T) {
 		ErrorChannel:   make(chan error),
 		MessageChannel: messageChannel,
 		Closer: func() error {
-			wg.Done()
+			defer wg.Done()
 			return nil
 		},
 	}

